CodegenProposal.md

```diff
@@ -0,0 +1,1474 @@
+
+Because `SelectionSet` field access uses unsafe force casting under the hood, it is necessary that a `SelectionSet` is only ever created via the execution process. A `SelectionSet` that is initialized manually cannot be guaranteed to contain all the expected data for its field accessors, and as such, could cause crashes at run time. `SelectionSet`s returned from GraphQL execution are guaranteed to be safe.
+
+## Nullable Arguments - `GraphQLNullable`
+---
+
+By default, `GraphQLOperation` field variables; fields on `InputObject`s; and field arguments are nullable. For a nullable argument, the value can be provided as a value, a `null` value, or omitted entirely. In `GraphQL`, omitting an argument and passing a `null` value have semantically different meanings. While often, they may be identical, it is up to the implementation of the server to interpret these values. For example, a `null` value for an argument on a mutation may indicate that a field on the object should be set to `null`, while omitting the argument indicates that the field should retain it's current value -- or be set to a default value.
+
+Because of the semantic difference between `null` and ommitted arguments, we have introduced `GraphQLNullable`. `GraphQLNullable` is a generic enum that acts very similarly to `Optional`, but it differentiates between a `nil` value (the `.none` case), and a `null` value (the `.null` case). Values are still wrapped using the `.some(value)` case as in `Optional`.
+
+The previous Apollo versions used double optionals (`??`) to represent `null` vs ` nil`. This was unclear to most users and make reading and reasoning about your code difficult in many situations. `GraphQLNullable` makes your intentions clear and explicit when dealing with nullable input values.
+
+For more information and implementation details, see: [GraphQLNullable.swift](Sources/ApolloAPI/GraphQLNullable.swift)
+
+# Generated Objects
+
+An overview of the format of all generated object types.
+
+# Schema Type Generation
+
+In addition to generating `SelectionSet`s for your `GraphQLOperation`s, types will be generated for each type (object, interface, or union) that is used in any operations across your entire application. These types will include all the fields that may be fetched by any operation used and can include other type metadata. 
+
+The schema types have a number of functions. 
+
+* Include metadata that allows the `GraphQLExecutor` and runtime type checking on `TypeCase`s to operate. 
+* Can be extended to provide cache key computation for types to configure the normalized cache.
+* Used for interacting with the cache for manual read/write functionality.
+* Used to create mock objects for generated `SelectionSet`s to be used in unit tests.
+
+
+These schema types can be included directly in your application target, or be generated into a separate shared library that can be used across modules in your application.
```

Comment:
I'm contemplating separating the type information needed for cache/validation from the type information needed for building mock objects. Since the later are to be used during tests, it would be nice not to embed them in the final app.

Reply:
Would be interested in hearing more about this, but it sounds like we are taking very different approaches to mock objects. I haven't started writing the code for my approach yet though, so I could be persuaded to take a different approach.